### PR TITLE
Adds private method to decode terminating null into nil [Fixes #173]

### LINKED ISSUE #173: Is there a concise way to decode a nested dictionary that may be null
Here is example JSON when the nested dictionary, "engine", is not null:

```
{
  "color" : "Red",
  "engine" : {
    "cylinder_count" : 8,
    "horsepower" : 300
  }
}
```

And here is example JSON when "engine" is null:

```
{
  "color" : "Red",
  "engine" : null
}
```

Here is the Engine model which works fine:

```
struct Engine: JSONDecodable {
  let cylinderCount: Int
  let horsepower: Int
  
  init(json: JSON) throws {
    cylinderCount = try json.int("cylinder_count")
    horsepower = try json.int("horsepower")
  }
}
```

And here is the Car model which also works fine but it seems like there should be a more concise way to handle the engine parsing:

```
struct Car: JSONDecodable {
  let color: String
  let engine: Engine?
  
  init(json: JSON) throws {
    color = try json.string("color")

    if let _ = try json.dictionary("engine", alongPath: [.NullBecomesNil]) {
      engine = try json.decode("engine", alongPath: [.NullBecomesNil])
    } else {
      engine = nil
    }
  }
}
```

I thought this would work:

```
struct Car: JSONDecodable {
  let color: String
  let engine: Engine?
  
  init(json: JSON) throws {
    color = try json.string("color")

    engine = try json.decode("engine", alongPath: [.NullBecomesNil])
  }
}
```

But it results in the error "Freddy.JSON.Error.UnexpectedSubscript(Swift.String)".

Am I missing something?  Is there a better way to handle this situation?  

Thank you very much, there are a lot of things I like about the framework!

### PR DESCRIPTION
**Pull request by mdmathias**
Users of Freddy had trouble decoding JSON that had a terminating Null into some JSONDecodable type. This commit adds a private static method to JSON that checks to see if the JSON in question is equal to .Null. If it is, then Error.ValueNotConvertible is thrown. Thus, if .NullBecomesNil along the path in question, then mapOptionalAtPath(_:alongPath:transform:) will return nil.

### CODE REVIEW

**Comment by mdmathias** [comment]
@zwaldowski Happy to have your 👀 as well.

File: Sources/JSONSubscripting.swift [hunk 1]
```diff
@@ -254,6 +254,17 @@ extension JSON {
 
 extension JSON {
 
+    /// Decodes a `JSON` instance if it is not `.Null`, throws otherwise.
+    /// - parameter json: An instance of `JSON`.
+    /// - returns: An instance of some type that conforms to `JSONDecodable`.
+    /// - throws: `Error.ValueNotConvertible` if the `JSON` instance is `.Null`.
+    private static func getDecoded<Decoded: JSONDecodable>(json: JSON) throws -> Decoded {
+        guard json != .Null else {
```
Comment (zwaldowski) [inline Sources/JSONSubscripting.swift]: I'd prefer `if case` here so as to not get into the `==` machinery, but that's just me.

File: Sources/JSONSubscripting.swift [hunk 2]
```diff
@@ -272,7 +283,7 @@ extension JSON {
     ///     instance does not match the decoded value.
     ///   * Any error that arises from decoding the value.
     public func decode<Decoded: JSONDecodable>(path: JSONPathType..., alongPath options: SubscriptingOptions, type: Decoded.Type = Decoded.self) throws -> Decoded? {
-        return try mapOptionalAtPath(path, alongPath: options, transform: Decoded.init)
+        return try mapOptionalAtPath(path, alongPath: options, transform: JSON.getDecoded)
```
Comment (zwaldowski) [inline Sources/JSONSubscripting.swift]: Seems like the `transform` could simply be elided as a trailing closure?

Comment (zwaldowski) [inline Sources/JSONSubscripting.swift]: Does the documentation of this method need to change accordingly?

**Comment by zwaldowski** [comment]
I'm fine with the change if it fixes an issue.

File: Sources/JSONSubscripting.swift (hunk 2 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: Re: eliding the `transform`, do you mean this?

```swift
public func decode<Decoded: JSONDecodable>(path: JSONPathType..., alongPath options: SubscriptingOptions, type: Decoded.Type = Decoded.self) throws -> Decoded? {
    return try mapOptionalAtPath(path, alongPath: options) { try JSON.getDecoded($0) }
}
```

I think passing `JSON.getDecoded` to `transform:` is more: a) parsimonious, and b) consistent with the code base's current approach.

Or, perhaps I'm not understanding your suggestion.

Comment (mdmathias) [inline Sources/JSONSubscripting.swift]: Re: changing the documentation...I don't think so.  I don't think that anything has changed from the caller's point of view.

File: Sources/JSONSubscripting.swift (hunk 1 above)
Comment (mdmathias) [inline Sources/JSONSubscripting.swift]:
```swift
if case .Null = json {
    throw Error.ValueNotConvertible(value: json, to: Decoded.self)
}
```

I'm ambivalent about this change.  If you think it's important, then I'll do it.  Let me know.  :)

**Pull request merged**